pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1895,6 +1895,24 @@ public void toggleQueryQuotaStateForBroker(String brokerInstanceName, String sta
     return serverToSegmentsMap;
   }
 
+  /**
+   * Returns a list of CONSUMING segments for the given realtime table.
+   */
+  public List<String> getConsumingSegments(String tableNameWithType) {
+    IdealState idealState = _helixAdmin.getResourceIdealState(_helixClusterName, tableNameWithType);
+    if (idealState == null) {
+      throw new IllegalStateException("Ideal state does not exist for table: " + tableNameWithType);
+    }
+    List<String> consumingSegments = new ArrayList<>();
+    for (String segment : idealState.getPartitionSet()) {
+      Map<String, String> instanceStateMap = idealState.getInstanceStateMap(segment);
+      if (instanceStateMap.containsValue(SegmentStateModel.CONSUMING)) {
```

Comment:
might also need null check?

Reply:
shouldn't need null check (or even Set) because we're iterating over the segments fetched from the ideal state. Changed it to Set anyway.